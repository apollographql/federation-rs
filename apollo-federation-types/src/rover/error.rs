use std::{
    error::Error,
    fmt::{self, Display},
};

use crate::build_plugin::{BuildMessage, BuildMessageLocation};
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct BuildError {
    /// A message describing the build error.
    pub message: Option<String>,

    /// A code describing the build error.
    pub code: Option<String>,

    /// The type of build error.
    #[serde(rename = "type")]
    error_type: BuildErrorType,

    /// Other untyped JSON included in the build output.
    #[serde(flatten)]
    other: crate::UncaughtJson,

    pub nodes: Option<Vec<BuildMessageLocation>>,

    omitted_nodes_count: Option<u32>,
}

impl From<BuildMessage> for BuildError {
    fn from(message: BuildMessage) -> Self {
        BuildError {
            message: Some(message.message),
            code: message.code,
            error_type: BuildErrorType::Composition,
            other: message.other,
            nodes: Some(message.locations),
            omitted_nodes_count: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct BuildErrorNodeLocation {
    subgraph: Option<String>,
}

impl BuildError {
    pub fn composition_error(
        code: Option<String>,
        message: Option<String>,
        nodes: Option<Vec<BuildMessageLocation>>,
        omitted_nodes_count: Option<u32>,
    ) -> BuildError {
        BuildError::new(
            code,
            message,
            BuildErrorType::Composition,
            nodes,
            omitted_nodes_count,
        )
    }

    pub fn config_error(code: Option<String>, message: Option<String>) -> BuildError {
        BuildError::new(code, message, BuildErrorType::Config, None, None)
    }

    fn new(
        code: Option<String>,
        message: Option<String>,
        error_type: BuildErrorType,
        nodes: Option<Vec<BuildMessageLocation>>,
        omitted_nodes_count: Option<u32>,
    ) -> BuildError {
        let real_message = if code.is_none() && message.is_none() {
            Some("An unknown error occurred during the build.".to_string())
        } else {
            message
        };
        BuildError {
            code,
            message: real_message,
            error_type,
            other: crate::UncaughtJson::new(),
            nodes,
            omitted_nodes_count,
        }
    }

    pub fn get_message(&self) -> Option<String> {
        self.message.clone()
    }

    pub fn get_code(&self) -> Option<String> {
        self.code.clone()
    }

    pub fn get_type(&self) -> BuildErrorType {
        self.error_type.clone()
    }

    pub fn get_nodes(&self) -> Option<Vec<BuildMessageLocation>> {
        self.nodes.clone()
    }

    pub fn get_omitted_nodes_count(&self) -> Option<u32> {
        self.omitted_nodes_count
    }
}

#[derive(Debug, Deserialize, Serialize, Clone, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
#[non_exhaustive]
pub enum BuildErrorType {
    Composition,
    Config,
}

impl Display for BuildError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            self.code.as_ref().map_or("UNKNOWN", String::as_str)
        )?;
        if let Some(message) = &self.message {
            write!(f, ": {message}")?;
        }
        Ok(())
    }
}

#[derive(Debug, Deserialize, Serialize, Default, Clone, PartialEq, Eq)]
pub struct BuildErrors {
    pub build_errors: Vec<BuildError>,

    #[serde(skip)]
    pub is_config: bool,
}

impl BuildErrors {
    pub fn new() -> Self {
        BuildErrors {
            build_errors: Vec::new(),
            is_config: false,
        }
    }

    pub fn iter(&self) -> impl Iterator<Item = &BuildError> {
        self.build_errors.iter()
    }

    pub fn len(&self) -> usize {
        self.build_errors.len()
    }

    pub fn length_string(&self) -> String {
        let num_failures = self.build_errors.len();
        if num_failures == 0 {
            unreachable!("No build errors were encountered while composing the supergraph.");
        }

        match num_failures {
            1 => "1 build error".to_string(),
            _ => format!("{num_failures} build errors"),
        }
    }

    pub fn push(&mut self, error: BuildError) {
        if matches!(error.error_type, BuildErrorType::Config) {
            self.is_config = true;
        }
        self.build_errors.push(error);
    }

    pub fn is_empty(&self) -> bool {
        self.build_errors.is_empty()
    }

    pub fn extend(&mut self, other: BuildErrors) {
        self.build_errors.extend(other.build_errors);
    }
}

impl Display for BuildErrors {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let num_failures = self.build_errors.len();
        if num_failures == 0
            || (num_failures == 1
                && self.build_errors[0].code.is_none()
                && self.build_errors[0].message.is_none())
        {
            writeln!(f, "Something went wrong! No build errors were recorded, but we also were unable to build a valid supergraph.")?;
        } else {
            for build_error in &self.build_errors {
                writeln!(f, "{build_error}")?;
            }
        }
        Ok(())
    }
}

#[cfg(feature = "config")]
impl From<crate::config::ConfigError> for BuildErrors {
    fn from(config_error: crate::config::ConfigError) -> Self {
        BuildErrors {
            build_errors: vec![BuildError::config_error(
                config_error.code(),
                Some(config_error.message()),
            )],
            is_config: true,
        }
    }
}

impl From<Vec<BuildError>> for BuildErrors {
    fn from(build_errors: Vec<BuildError>) -> Self {
        let is_config = build_errors
            .iter()
            .any(|e| matches!(e.error_type, BuildErrorType::Config));
        BuildErrors {
            build_errors,
            is_config,
        }
    }
}

impl FromIterator<BuildError> for BuildErrors {
    fn from_iter<I: IntoIterator<Item = BuildError>>(iter: I) -> Self {
        let mut c = BuildErrors::new();

        for i in iter {
            c.push(i);
        }

        c
    }
}

impl IntoIterator for BuildErrors {
    type Item = BuildError;
    type IntoIter = std::vec::IntoIter<BuildError>;

    fn into_iter(self) -> Self::IntoIter {
        self.build_errors.into_iter()
    }
}

impl Error for BuildError {}
impl Error for BuildErrors {}

#[cfg(test)]
mod tests {
    use super::{BuildError, BuildErrors};

    use crate::build_plugin::BuildMessageLocation;
    use serde_json::{json, Value};

    #[test]
    fn it_supports_iter() {
        let build_errors: BuildErrors = vec![
            BuildError::composition_error(None, Some("wow".to_string()), None, None),
            BuildError::composition_error(
                Some("BOO".to_string()),
                Some("boo".to_string()),
                None,
                None,
            ),
        ]
        .into();

        let messages: Vec<String> = build_errors
            .iter()
            .map(|e| e.get_message().unwrap())
            .collect();

        assert_eq!(messages, vec!["wow", "boo"]);
    }

    #[test]
    fn it_can_serialize_empty_errors() {
        let build_errors = BuildErrors::new();
        assert_eq!(
            serde_json::to_string(&build_errors).expect("Could not serialize build errors"),
            json!({"build_errors": []}).to_string()
        );
    }

    #[test]
    fn it_can_serialize_some_build_errors() {
        let error_node = BuildMessageLocation {
            subgraph: Some("foo".to_string()),
            ..Default::default()
        };

        let build_errors: BuildErrors = vec![
            BuildError::composition_error(
                None,
                Some("wow".to_string()),
                Some(vec![error_node.clone()]),
                Some(1),
            ),
            BuildError::composition_error(
                Some("BOO".to_string()),
                Some("boo".to_string()),
                Some(vec![error_node.clone()]),
                Some(2),
            ),
        ]
        .into();

        let actual_value: Value = serde_json::from_str(
            &serde_json::to_string(&build_errors)
                .expect("Could not convert build errors to string"),
        )
        .expect("Could not convert build error string to serde_json::Value");

        let expected_value = json!({
            "build_errors": [
              {
                "message": "wow",
                "code": null,
                "type": "composition",
                "nodes": [
                  {
                      "subgraph": "foo",
                      "source": null,
                      "start": null,
                      "end": null
                  }
                ],
                "omittedNodesCount": 1
              },
              {
                "message": "boo",
                "code": "BOO",
                "type": "composition",
                "nodes": [
                  {
                      "subgraph": "foo",
                      "source": null,
                      "start": null,
                      "end": null
                  }
                ],
                "omittedNodesCount": 2
              }
            ]
        });
        assert_eq!(actual_value, expected_value);
    }

    #[test]
    fn it_can_deserialize() {
        let msg = "wow".to_string();
        let code = "boo".to_string();
        let actual_struct = serde_json::from_str(
            &json!({ "message": &msg, "code": &code, "type": "composition", "nodes": null, "omittedNodesCount": 12 }).to_string(),
        ).unwrap();
        let expected_struct =
            BuildError::composition_error(Some(code.clone()), Some(msg.clone()), None, Some(12));
        assert_eq!(expected_struct, actual_struct);
    }
}
