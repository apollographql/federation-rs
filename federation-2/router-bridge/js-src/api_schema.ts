import {
  buildSchema as buildGraphqlSchema,
  ExecutionResult,
  GraphQLError,
  printSchema,
} from "graphql";

import { buildSupergraphSchema } from "@apollo/federation-internals";

export function apiSchema(sdl: string): ExecutionResult<String> {
  let schema: String;
  try {
    // First go through regular schema parsing
    buildGraphqlSchema(sdl);

    // Now try to get the API schema
    let [composedSchema] = buildSupergraphSchema(sdl);

    let apiSchema = composedSchema.toAPISchema();
    schema = printSchema(apiSchema.toGraphQLJSSchema());
  } catch (e) {
    return {
      errors: [e],
    };
  }
  if (!schema) {
    return {
      errors: [new GraphQLError("couldn't build api schema from SDL")],
    };
  }
  return { data: schema, errors: [] };
}
