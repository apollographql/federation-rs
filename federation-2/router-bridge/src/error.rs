/*!
# Errors raised by the `router-bridge` when trying to run `javascript`.
*/

use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Error, Serialize, Deserialize, Debug, Clone)]
/// An error which occurred within the bridge.
///
/// This does not include JS domain related errors, such as [`GraphQLError`].
pub enum Error {
    /// An uncaught error was raised when invoking a custom script.
    ///
    /// This contains the script invocation error message.
    #[error("the deno runtime raised an error: `{0}`")]
    DenoRuntime(String),
    /// An uncaught error was raised when trying to serialize a parameter before sending it to the javascript worker.
    ///
    /// This contains the serialization error message, and the payload name.
    #[error("couldn't serialize parameter `{name}`: `{message}`.")]
    ParameterSerialization {
        /// The underlying serialization error.
        message: String,
        /// The name of the parameter we tried to serialize.
        name: String,
    },

    /// An uncaught error was raised when trying to deserialize a payload.
    ///
    /// This contains the deserialization error message, and the payload.
    #[error("couldn't deserialize payload `{id}`: `{message}`.")]
    ParameterDeserialization {
        /// The underlying serialization error.
        message: String,
        /// The deno response id we tried to deserialize.
        id: String,
    },
}
