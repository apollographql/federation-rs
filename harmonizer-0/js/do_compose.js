/** @typedef {{typeDefs: string, name: string, url?: string;}} ServiceDefinition */

/**
 * This `composition` is defined as a global by the runtime we define in Rust.
 * We declare this as a `var` here only to allow the TSDoc type annotation to be
 * applied to it. Running `var` multiple times has no effect.
 * @type {{
 *   composeAndValidate: import('../../federation-js').composeAndValidate,
 *   parseGraphqlDocument: import('graphql').parse
 * }} */
var composition;

/**
 * @type {ServiceDefinition[]}
 */
var serviceList = serviceList;

if (!serviceList || !Array.isArray(serviceList)) {
  throw new Error("Error in JS-Rust-land: serviceList missing or incorrect.");
}

serviceList.some((service) => {
  if (
    typeof service.name !== 'string' ||
    !service.name ||
    (typeof service.url !== 'string' && service.url) ||
    (typeof service.sdl !== 'string' && service.sdl)
  ) {
    throw new Error("Missing required data structure on service.");
  }
});

serviceList = serviceList.map(({ sdl, ...rest }) => ({
  typeDefs: parseTypedefs(sdl),
  ...rest,
}));

function parseTypedefs(source) {
  try {
    return composition.parseGraphqlDocument(source)    
  } catch (err) {
    // Return the error in a way that we know how to handle it.
    done({ Err: [err] });
  }
}

try {
  /**
   * @type {{ errors: Error[], supergraphSdl?: undefined } | { errors?: undefined, supergraphSdl: string; }}
   */
  const composed = composition.composeAndValidate(serviceList);
  done(
    composed.errors ? { Err: composed.errors } : { Ok: composed.supergraphSdl },
  );
} catch (err) {
  done({ Err: [err] });
}
