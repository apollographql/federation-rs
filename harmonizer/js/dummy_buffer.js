dummy_buffer = require("../node_modules/buffer").Buffer;
