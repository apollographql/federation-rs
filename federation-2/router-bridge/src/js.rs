use crate::error::Error;
/// Wraps creating the Deno Js runtime collecting parameters and executing a script.
use deno_core::{
    anyhow::{anyhow, Error as AnyError},
    op, Extension, JsRuntime, OpState, RuntimeOptions, Snapshot,
};
use serde::de::DeserializeOwned;
use serde::Serialize;
use std::sync::mpsc::{channel, Sender};

// A reasonable default starting limit for our deno heap.
const APOLLO_ROUTER_BRIDGE_EXPERIMENTAL_V8_INITIAL_HEAP_SIZE_DEFAULT: &str = "256";

pub(crate) struct Js {
    name: String,
    parameters: Vec<(&'static str, String)>,
}

impl Js {
    pub(crate) fn new(name: String) -> Js {
        Js {
            name,
            parameters: Vec::new(),
        }
    }

    pub(crate) fn with_parameter<T: Serialize>(
        mut self,
        name: &'static str,
        param: T,
    ) -> Result<Js, Error> {
        let serialized = format!(
            "{} = {}",
            name,
            serde_json::to_string(&param).map_err(|error| Error::ParameterSerialization {
                name: name.to_string(),
                message: error.to_string()
            })?
        );
        self.parameters.push((name, serialized));
        Ok(self)
    }

    pub(crate) fn execute<Ok: DeserializeOwned + 'static>(
        &self,
        name: &'static str,
        source: &'static str,
    ) -> Result<Ok, Error> {
        // We'll use this channel to get the results
        let (tx, rx) = channel::<Result<Ok, Error>>();

        let happy_tx = tx.clone();

        let my_ext = Extension::builder()
            .ops(vec![deno_result::decl::<Ok>()])
            .state(move |state| {
                state.put(happy_tx.clone());
                Ok(())
            })
            .build();

        let mut runtime = self.build_js_runtime(my_ext);

        for parameter in self.parameters.iter() {
            runtime
                .execute_script(format!("<{}>", parameter.0).as_str(), &parameter.1)
                .expect("unable to evaluate service list in JavaScript runtime");
        }

        // We are sending the error through the channel already
        let _ = runtime.execute_script(name, source).map_err(|e| {
            let message = format!(
                "unable to invoke `{name}` in JavaScript runtime \n error: \n {:?}",
                e
            );

            tx.send(Err(Error::DenoRuntime(message)))
                .expect("channel must be open");

            e
        });

        rx.recv().expect("channel remains open")
    }

    pub(crate) fn build_js_runtime(&self, my_ext: Extension) -> JsRuntime {
        // Initialize a runtime instance
        let buffer = include_bytes!(concat!(env!("OUT_DIR"), "/query_runtime.snap"));

        let heap_size =
            match std::env::var("APOLLO_ROUTER_BRIDGE_EXPERIMENTAL_V8_INITIAL_HEAP_SIZE") {
                Ok(v) => v,
                Err(_e) => {
                    APOLLO_ROUTER_BRIDGE_EXPERIMENTAL_V8_INITIAL_HEAP_SIZE_DEFAULT.to_string()
                }
            };

        // The first flag is argv[0], so provide an ignorable value
        let flags = vec![
            "--ignored".to_string(),
            "--max-heap-size".to_string(),
            heap_size,
        ];

        // Deno will warn us if we supply flags it doesn't recognise.
        // We ignore "--ignored" and report any others as warnings
        let ignored: Vec<_> = deno_core::v8_set_flags(flags)
            .into_iter()
            .filter(|x| x != "--ignored")
            .collect();
        if !ignored.is_empty() {
            tracing::warn!("deno ignored these flags: {:?}", ignored);
        }

        let mut js_runtime = JsRuntime::new(RuntimeOptions {
            extensions: vec![my_ext],
            startup_snapshot: Some(Snapshot::Static(buffer)),
            ..Default::default()
        });

        // Add a callback that expands our heap by 1.25 each time
        // it is invoked. There is no limit, since we rely on the
        // execution environment (OS) to provide that.
        let name = self.name.clone();
        js_runtime.add_near_heap_limit_callback(move |current, initial| {
            let new = current * 5 / 4;
            tracing::info!(
                "deno heap expansion({}): initial: {}, current: {}, new: {}",
                name,
                initial,
                current,
                new
            );
            new
        });
        js_runtime
    }
}

#[op]
fn deno_result<Response>(state: &mut OpState, payload: Response) -> Result<(), AnyError>
where
    Response: DeserializeOwned + 'static,
{
    // we're cloning here because we don't wanna keep the borrow across an await point
    let sender = state.borrow::<Sender<Result<Response, Error>>>().clone();
    sender
        .send(Ok(payload))
        .map_err(|e| anyhow!("couldn't send response {e}"))
}
