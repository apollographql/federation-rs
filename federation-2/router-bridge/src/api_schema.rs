/*!
# Generate an API schema from an sdl.
*/

use crate::error::Error;
use crate::js::Js;
use serde::{Deserialize, Serialize};
use std::fmt::Display;
use thiserror::Error;

/// An error which occurred during JavaScript api schema generation.
///
/// The shape of this error is meant to mimick that of the error created within
/// JavaScript, which is a [`GraphQLError`] from the [`graphql-js`] library.
///
/// [`graphql-js']: https://npm.im/graphql
/// [`GraphQLError`]: https://github.com/graphql/graphql-js/blob/3869211/src/error/GraphQLError.js#L18-L75
#[derive(Debug, Error, Serialize, Deserialize, PartialEq, Eq, Clone)]
pub struct ApiSchemaError {
    /// A human-readable description of the error that prevented api schema generation.
    pub message: Option<String>,
}

impl Display for ApiSchemaError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(self.message.as_deref().unwrap_or("UNKNOWN"))
    }
}

/// The type returned when invoking `api_schema`
pub type ApiSchemaResult = Result<String, Vec<ApiSchemaError>>;

/// The `api_schema` function receives a [`string`] representing the SDL and invokes JavaScript
/// functions to parse, convert to apiSchema and print to string.
///
pub fn api_schema(sdl: &str) -> Result<ApiSchemaResult, Error> {
    Js::new()
        .with_parameter("sdl", sdl)?
        .execute::<ApiSchemaResult>("do_api_schema", include_str!("../bundled/do_api_schema.js"))
}

#[cfg(test)]
mod tests {
    use crate::api_schema::{api_schema, ApiSchemaError};

    #[test]
    fn it_works() {
        let raw_sdl = include_str!("testdata/contract_schema.graphql");

        let api_schema = api_schema(raw_sdl).unwrap();
        insta::assert_snapshot!(&api_schema.unwrap());
    }

    #[test]
    fn invalid_sdl() {
        let expected_error = ApiSchemaError {
            message: Some(r#"Unknown type "Query"."#.to_string()),
        };
        let response = api_schema(
            "schema {
                query: Query
            }",
        )
        .expect("an uncaught deno error occured");

        assert_eq!(response.err().unwrap(), vec![expected_error]);
    }
}
