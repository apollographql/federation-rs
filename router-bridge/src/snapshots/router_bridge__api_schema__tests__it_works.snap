---
source: router-bridge/src/api_schema.rs
assertion_line: 60
expression: "&api_schema.unwrap()"

---
directive @apollo_studio_metadata(launchId: String, buildId: String, checkId: String) on SCHEMA

type Mutation {
  createProduct(name: String, upc: ID!): Product
  createReview(body: String, id: ID!, upc: ID!): Review
}

type Product {
  name: String
  price: Int
  reviews: [Review]
  reviewsForAuthor(authorID: ID!): [Review]
  shippingEstimate: Int
  upc: String!
  weight: Int
}

type Query {
  me: User
  topProducts(first: Int = 5): [Product]
}

type Review {
  author: User
  body: String
  id: ID!
  product: Product
}

type User {
  id: ID!
  name: String
  reviews: [Review]
  username: String
}
